ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/recovery/arjunacore/XARecoveryModule.java

```diff
@@ -968,9 +995,9 @@ private boolean isHelperInUse(XAResourceRecoveryHelper xaResourceRecoveryHelper)
      */
     private boolean waitForScanState(ScanStates state) {
         try {
-            do {
+			while (!getScanState().equals(state)) {
```

Comment:
why did you change the do ... while into a while do loop. The caller already knows the scan state is not equal to state and since this is a private method it is safe to assume that the caller has already checked (under the lock) that he has to wait for a state change.